contracts/external/cw-token-swap/src/types.rs

```diff
@@ -0,0 +1,420 @@
+                amount,
+                WasmMsg::Execute {
+                    contract_addr: cw20_addr.to_string(),
+                    msg: to_binary(&cw20::Cw20ExecuteMsg::Transfer {
+                        recipient: deps.api.addr_validate(&recipient)?.to_string(),
+                        amount,
+                    })?,
+                    funds: vec![],
+                }
+                .into(),
+            )),
+        }
+    }
+}
+
+/// Information about a counterparty in this escrow transaction and
+/// their promised funds.
+#[cw_serde]
+pub struct Counterparty {
+    /// The address of the counterparty.
+    pub address: String,
+    /// The funds they have promised to provide.
+    pub promise: SwapInfo,
+}
+
+impl Counterparty {
+    pub fn into_checked(self, deps: Deps) -> Result<CheckedCounterparty, ContractError> {
+        Ok(CheckedCounterparty {
+            address: deps.api.addr_validate(&self.address)?,
+            provided: false,
+            promise: self.promise.into_checked(deps)?,
+        })
+    }
+}
+
+#[cw_serde]
+pub struct CheckedCounterparty {
+    pub address: Addr,
+    pub promise: CheckedSwapInfo,
+    pub provided: bool,
+}
+
+/// Information about the token being used on one side of the escrow.
+#[cw_serde]
+pub enum SwapInfo {
+    /// A native token.
+    Native {
+        denom: String,
+        amount: Uint128,
+        on_completion: Vec<NativeSendMsgs>,
+    },
+    /// A cw20 token.
+    Cw20 {
+        contract_addr: String,
+        amount: Uint128,
+        on_completion: Vec<Cw20SendMsgs>,
+    },
+}
+
+impl SwapInfo {
+    pub fn into_checked(self, deps: Deps) -> Result<CheckedSwapInfo, ContractError> {
+        match self {
+            SwapInfo::Native {
+                denom,
+                amount,
+                on_completion,
+            } => {
+                if amount.is_zero() {
+                    Err(ContractError::ZeroTokens {})
+                } else {
+                    let on_completion = if on_completion.is_empty() {
+                        vec![]
+                    } else {
+                        let mut total_amount = Uint128::zero();
+                        let cosmos_msgs = on_completion
+                            .into_iter()
+                            .map(|msg| {
+                                let (amount, cosmos_msg) =
+                                    msg.into_checked_cosmos_msg(deps, &denom)?;
+                                total_amount += amount;
+                                Ok(cosmos_msg)
+                            })
+                            .collect::<Result<Vec<CosmosMsg>, ContractError>>()?;
+
+                        // Verify that total amount of funds matches funds sent in all messages
+                        if total_amount != amount {
+                            return Err(ContractError::WrongFundsCalculation {});
+                        }
+                        cosmos_msgs
+                    };
```

Comment:
reading this code i my spidey sense was tingling. first, the iteration here can be done functionally with a fold over a tuple (no need for `mut` _and_ `map`):

```rust
                    let (on_completion, tokens_sent) = on_completion.into_iter().try_fold(
                        (vec![], Uint128::zero()),
                        |(mut messages, total_sent), msg| -> Result<_, ContractError> {
                            let (sent, msg) = msg.into_checked_cosmos_msg(deps, &denom)?;
                            messages.push(msg);
                            Ok((messages, sent + total_sent))
                        },
                    )?;
```

second, only needing to check the length in the `on_completion.len() > 0` case seems strange. the semantics of that suggest that nothing will be done on completion meaning... what? there exist both a bank and cw20 transfer option in the message types, so shouldn't a transfer to the other counterparty be represented as one of those? why have both ways to trigger a transfer to the other counterparty?

this is one of those weird corner cases of code where the logic gets weird and you realize you've got a wrong abstraction. how about making doing nothing on completion not allowed, and expressing the desire to transfer tokens with the transfer completion options?

Reply:
1. Thats a neater way of doing it, I like it.
2. I look at it as having a default, at the end of the day we are talking about a swap contract, the default behavior is sending the tokens to the other party, this design allows having a basic swap in place without adding complexity to calling it, so providing an empty vector will result in a default swap behavior.
If you need something more complex, you can provide the msgs you want in the `on_completion` field.

Im not against forcing all msgs on `on_completion`, I just think having this default makes a much better UI.

Thinking about it, I think I can set the default here instead of later, which will make it much clearer that we set default.

Reply:
(2) sounds like a good idea. the default can be to fill with a transfer message, maybe we just do that at instantiation time?

Reply:
might be nice to break this logic into a function as well so it can be reused in the cw20 case